sql: don't copy old pk when changing bucket size on sharded index

rohany opened the pull request:
Fixes #45889.

Release note (sql change): This PR causes primary key changes to
not create a copy of the old primary key if the primary key change
only changes the bucket count of a hash sharded index.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45894)
<!-- Reviewable:end -->

otan commented:
Just for sanity - why do we not want this behaviour?

rohany commented:
The argument is as follows:

We make a copy of the old primary key so that your query performance doesn't regress after the primary key change. However, in the hash sharded case, the change being made is going to affect performance (i.e. what is the bucket count, or what columns are hashed). Keeping around the old primary key means keeping around the old primary key with the incorrect number of buckets etc. What does @ajwerner think about this?

ajwerner reviewed (commented):
I agree with @rohany here, at least for the tested case. I can't see any reason why a user would want to retain the old bucket count.

Is there are case where the user would want to retain the index if they changed the set of columns but both were hashed? That one is less clear to me.


Consider:

```
CREATE TABLE t (x INT NOT NULL, y INT NOT NULL, ts TIMESTAMPTZ NOT NULL, PRIMARY KEY (ts, x, y) USING HASH WITH BUCKET_COUNT = 2);
```
then:

```
ALTER TABLE t ALTER PRIMARY KEY USING COLUMNS (ts, y, x) USING HASH WITH BUCKET_COUNT=3
```

I could see it either way in this case. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45894#-:-M2-fszm9i-iuJ_-Iy0I:b-7pn27k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)



<!-- Sent from Reviewable.io -->

otan commented on pkg/sql/alter_table.go:
```diff
@@ -563,7 +563,9 @@ func (n *alterTableNode) startExec(params runParams) error {
 			// does, but doesn't store anything. We only recreate the index if
 			// the table has a primary key (no DROP PRIMARY KEY statements have
 			// been executed), and the primary key is not the default rowid key.
```
nit: maybe updated comment for hash sharded cases too

jordanlewis commented:
Does this want to be merged?

rohany committed:
sql: don't copy old pk when changing bucket size on sharded index

Fixes #45889.

Release note (sql change): This PR causes primary key changes to
not create a copy of the old primary key if the primary key change
only changes the bucket count of a hash sharded index.

rohany commented:
Reviving this PR! I am more satisfied with it now -- I changed it to only not create a copy if only the bucket count of the hash sharded index is changing. PTAL

otan commented on pkg/sql/alter_primary_key.go:
```diff
@@ -327,3 +325,20 @@ func (p *planner) AlterPrimaryKey(
+	oldPK := desc.PrimaryIndex
+	return desc.HasPrimaryKey() && !desc.IsPrimaryIndexDefaultRowID() &&
+		!(oldPK.IsSharded() && newPK.IsSharded() &&
+			// The first column in the columnIDs is the shard column, which will be different.
```
this was a teensy bit hard to read but makes sense after squinting. 

wondering if breaking down to the `if` statements might be better... but it's personal taste for me.

````

func shouldCopyPrimaryKey(desc *MutableTableDescriptor, newPK *sqlbase.IndexDescriptor) bool {
  // no primary key means we have nothing to copy
  if !desc.HasPrimaryKey() {
     return false
  }
  // we never want to copy default row id
  if desc.IsPrimaryIndexDefaultRowID() {
    return false
  }
  if oldPK.IsSharded() && newPK.IsSharded() {
     // ...
  }
  return true
}
```

rohany replied:
I'm down for this, I find it hard to read as well.

otan reviewed (commented):
LGTM. Completely optional bikeshedding comment.

rohany commented:
bors r=otan
